---
title: "Introduction to `MotifFunc`"
author: "Minh An Ho"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MotifFuncVignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = ">",
  fig.width=6, fig.height=6
)
```

# Functionality Visualization for Motifs

## Introduction

MotifFunc contains functions that handle genomic data to classify motifs, determine functionality, and group broad functionalities for word cloud visualization. As with the word cloud format, functions with the same frequency are displayed in the same colour and size.

## Installation

You can install the MotifFunc from github with:

``` 
require("devtools")
devtools::install_github("minhanho/MotifFunc", build_vignettes = TRUE)
library("MotifFunc")
```

You can also run the package by downloading the repository, followed by the commands:
``` 
setwd("<package path>")
devtools::load_all()
``` 

To run the shiny app:

``` 
runMotifFunc()
``` 

## Licensing 
`MotifFunc` is licensed under the MIT license. If you use `MotifFunc` in published research, please cite:

Ho, M. (2019). MotifFunc R. Retrieved from https://github.com/minhanho/MotifFunc (Original work published 2019)

## Getting Help
Create an issue on github https://github.com/minhanho/MotifFunc/issues


## Example Sequence of Functions
```
PWMfile <- system.file("extdata", "MA0007.1.transfac", package = "MotifFunc")
jaspar.scores <- MotifFunc:::jaspar.scores
matchNames <- MotifFunc::classifyPcmMotifs(PWMfile)
functionFreq <- MotifFunc::(matchNames)
```
Or

```
jaspar.scores <- MotifFunc:::jaspar.scores
matchNames <- MotifFunc::classifySeqMotifs("AGCGTAGGCGT")
functionFreq <- MotifFunc::getFunctionWC(matchNames)
```

## Retreiving an Input Data Set
When using `classifySeqMotifs()`, the input is a string composed of your choice of nucleotides (i.e. "A", "C", "G", "T")

But when using `classifyPcmMotifs()`, you must specify a path on your computer to a transfac or txt file containing frequency matrix data, to do this:
* Navigate to http://jaspar.genereg.net/
* Select a profile
* Download the "Frequency Matrix" in TRANSFAC format

## Helper Functions
A number of non-exported helper functions can be found in package repository to aid in various data processing tasks.
* `correctJasparTransfac()` - processes .transfac files and produces a .txt file that is compatible with `MotIV::readPWMfile()` to use as a query motif(s)
* `MotIV.toTable()` - returns list of match information from S4 motiv object
* `getFullOrganism()` - returns full name(s) of abbreviated organism names from `MotifDb` motif matches for use with `biomartr` functions

## BioMart and ENSEMBL Dependencies

The Wordcloud visualization and thus shiny app is directly dependent on the BioMart server. Without internet connection or in the case of BioMart server issues, the wordcloud will not be produced.
